drivers/adc/adc_cc23x0.c

```
@@ -0,0 +1,435 @@
+	case 8:
+		return ADC_CLOCK_DIVIDER_8;
+	case 16:
+		return ADC_CLOCK_DIVIDER_16;
+	case 24:
+		return ADC_CLOCK_DIVIDER_24;
+	case 32:
+		return ADC_CLOCK_DIVIDER_32;
+	case 48:
+		return ADC_CLOCK_DIVIDER_48;
+	default:
+		return ADC_CLOCK_DIVIDER_1;
+	}
+}
+
+static int adc_cc23x0_calc_clk_cfg(uint32_t acq_time_ns, uint32_t *clk_div, uint16_t *clk_cycles)
+{
+	const uint32_t base_clock = CPU_FREQ;
+	uint32_t divider;
+	uint32_t clock_frequency;
+	uint16_t min_cycles;
+	uint16_t cycles;
+	float clock_period_ns;
+
+	/* Initialize best values with maximum possible cycles */
+	min_cycles = ADC_CC23_MAX_CYCLES;
+	*clk_div = 0;
+	*clk_cycles = 0;
+
+	LOG_DBG("Requested sample duration: %u ns", acq_time_ns);
+
+	/* Iterate through each divider */
+	ARRAY_FOR_EACH(clk_dividers, i) {
+		divider = clk_dividers[i];
+		clock_frequency = base_clock / divider;
+
+		/* Calculate the clock period */
+		clock_period_ns = 1.0E9 / clock_frequency;
+
+		/* Calculate the number of cycles needed to meet or exceed acq_time_ns */
+		cycles = DIV_ROUND_UP(acq_time_ns, clock_period_ns);
+
+		/* Check if this configuration is valid and optimal */
+		if (cycles <= ADC_CC23_MAX_CYCLES && cycles < min_cycles) {
+			min_cycles = cycles;
+			*clk_div = divider;
+			*clk_cycles = cycles;
+
+			LOG_DBG("Divider: %u, Cycles: %u, Actual sample duration: %u ns",
+				divider, cycles, (uint32_t)clock_period_ns * cycles);
+		}
+	}
+
+	/* Check if a valid configuration was found */
+	if (*clk_div == 0) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int adc_cc23x0_channel_setup(const struct device *dev,
+				    const struct adc_channel_cfg *channel_cfg)
+{
+	struct adc_cc23x0_data *data = dev->data;
+	const uint8_t ch = channel_cfg->channel_id;
+	uint32_t ref;
+	uint32_t acq_time_ns;
+	uint32_t clk_div;
+	uint16_t clk_cycles;
+	int ret;
+
+	LOG_DBG("Channel %d", ch);
+
+	if (data->curr_mem_index > ADC_CC23_MEM_MAX) {
+		LOG_ERR("%d channels are already configured", ADC_CC23_MEM_COUNT);
```

Is there anyway to un-setup a channel so the memctl register can be freed up to be used for a new channel? Or can you only set up 4 channels, and then you are stuck with those 4 channels forever?

---

One way to support more channels is to maintain a list of memctl configs in RAM for each configured channel and then in `adc_cc23x0_read_common()` (which I haven't fully reviewed yet) we can copy the relevant memctl configs from RAM to the HW. And if more than 4 channels are requested in `adc_cc23x0_read_common()` then it could be split into multiple sequences.

---

When is re-initialization necessary? 

I now realize that the driverlib API doesn't configure the memctl register with just one call. If we want to use the current driverlib implementation then `adc_cc23x0_channel_setup()` would need to store the following information, in `data`, for the channel to be setup (maybe I am missing some):
 - reference
 - channel number
 - adjustment gain (which you already store)
 - sample cycles
 - clock divider

This could be implemented as an array, in `adc_cc23x0_data `, of 16 elements of a struct with members for each of above values.
 
Then in `adc_cc23x0_read_common()` you will call the relevant driverlib APIs to configure the memctl register(s) for the channels to be read before starting the conversion. If applicable you would do this (configure memctl registers and wait for conversion to complete) in a loop until all channels have been read. For example if more than 4 channels are to be read you can read 4 at a time until all channels have been read. Also, if some channels are not compatible to be read in one "sequence", for example if they require different clock dividers, then it could be done in two different iterations of the loop as well.

---

That might be an interesting feature indeed. But the goal of this PR is mainly to get a clean basis upstream.

---

I would argue that it is not just a nice feature, but necessary to comply with the Zephyr API. Currently the driver doesn't support the following:
 - More than 4 channels
 - Different references for different channels 
 - Different sample times for different channels
 - Different adjustment gains for different channels

The first one could possibly be an acceptable limitation since you the API would return an error, but for the others we would just silently use incorrect configurations.

---

The PR has been updated since your first review:
- Different references for different channels is now possible
- Different adjustment gains for different channels is now possible

For this one:
- Different sample times for different channels
...I agree with you, I can add an error (or a warning). I'll do.